File: GobusKMM/gobusIosApp/gobusIosApp/MethodChannelController.swift
//
//  MethodChannelController.swift
//  gobusIosApp
//
//  Created by Jesus Daniel Medina on 02/02/21.
//  Copyright © 2021 orgName. All rights reserved.
//
import Flutter
import gobusShared

class MethodChannelController {
  let binaryMessenger: FlutterBinaryMessenger
  
  init(_ binaryMessenger: FlutterBinaryMessenger) {
    self.binaryMessenger = binaryMessenger
  }
  
  func createUseCaseChannel() {
    let array = [
      MethodHandler.IncreaseCounterMethodChannel.init(increaseCounterUseCase: IncreaseCounterUseCaseImpl.init())
    ]
    MethodChannel.UseCaseChannel.init(methodHandlers: array).createFlutterMethodChannel(binaryMessenger)
  }
}

extension MethodChannel.UseCaseChannel {
  func createFlutterMethodChannel(_ binaryMessenger: FlutterBinaryMessenger) {
    FlutterMethodChannel(
      name: channelName,
      binaryMessenger: binaryMessenger
    ).setMethodCallHandler({
      (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
      let methodHandler = self.methodHandlers.first(where: { $0.methodName == call.method })
      switch (methodHandler) {
      case is MethodHandler.IncreaseCounterMethodChannel:
        let increaseCounterUseCase = (methodHandler as! MethodHandler.IncreaseCounterMethodChannel).increaseCounterUseCase
        let response = increaseCounterUseCase.invoke()
        result(response)
        break;
      default:
        result(FlutterMethodNotImplemented)
        break;
      }
    })
  }
}

File: GobusKMM/gobusAndroidApp/src/main/java/com/mupper/gobus/gobusAndroidApp/MethodChannelController.kt
package com.mupper.gobus.gobusAndroidApp

import com.mupper.gobus.gobusShared.IncreaseCounterUseCaseImpl
import com.mupper.gobus.gobusShared.MethodChannel.UseCaseChannel
import com.mupper.gobus.gobusShared.MethodHandler
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel


/**
 * Created by jesus.medina on 01/2021.
 * Insulet Corporation
 * Andromeda
 */
class MethodChannelController(private val flutterEngine: FlutterEngine) {
    fun createUseCaseMethodChannel() {
        UseCaseChannel(listOf(MethodHandler.IncreaseCounterMethodChannel(IncreaseCounterUseCaseImpl()))).createFlutterMethodChannel()
    }

    private fun UseCaseChannel.createFlutterMethodChannel() {
        MethodChannel(
            flutterEngine.dartExecutor.binaryMessenger,
            channelName
        ).setMethodCallHandler { call, result ->
            when (val methodHandler = methodHandlers.first { it.methodName == call.method }) {
                is MethodHandler.IncreaseCounterMethodChannel -> {
                    val answer = methodHandler.increaseCounterUseCase()
                    result.success(answer)
                }
                else -> result.notImplemented()
            }
        }
    }
}